vsintegration/src/FSharp.Editor/LanguageService/UntypedAstUtils.fs

```diff
@@ -0,0 +1,417 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information
+namespace Microsoft.VisualStudio.FSharp.Editor
+open System
+open System.Collections.Generic
+
+open Microsoft.FSharp.Compiler
+open Microsoft.FSharp.Compiler.Ast
+
+[<AutoOpen>]
+module internal UntypedAstUtils =
+    open Microsoft.FSharp.Compiler.Range
+
+    type Range.range with
+        member inline x.IsEmpty = x.StartColumn = x.EndColumn && x.StartLine = x.EndLine 
+
+    type internal ShortIdent = string
+    type internal Idents = ShortIdent[]
+
+    let internal longIdentToArray (longIdent: Ident list): Idents =
+        longIdent |> Seq.map string |> Seq.toArray   
+
+
+    /// An recursive pattern that collect all sequential expressions to avoid StackOverflowException
+    let rec (|Sequentials|_|) = function
+        | SynExpr.Sequential(_, _, e, Sequentials es, _) ->
+            Some(e::es)
+        | SynExpr.Sequential(_, _, e1, e2, _) ->
+            Some [e1; e2]
+        | _ -> None
+
+    let (|ConstructorPats|) = function
+        | SynConstructorArgs.Pats ps -> ps
+        | SynConstructorArgs.NamePatPairs(xs, _) -> List.map snd xs
+
+
+    /// Returns all Idents and LongIdents found in an untyped AST.
+    let internal getLongIdents (input: ParsedInput option) : IDictionary<Range.pos, Idents> =
```

Comment:
I already try to port all signature generation code to NicePrint.fs . 

Thanks for referring to this.